src/access/Subjects.h

```diff
@@ -0,0 +1,208 @@
+        else
+        {
+            return Access::AuthMode::kNone;
+        }
+    }
+
+#if CHIP_DETAIL_LOGGING
+    const char * GetAuthModeString() const
+    {
+        switch (GetAuthMode())
+        {
+        case Access::AuthMode::kPase:
+            return "Pase";
+        case Access::AuthMode::kCase:
+            return "Case";
+        case Access::AuthMode::kGroup:
+            return "Group";
+        case Access::AuthMode::kNone:
+        default:
+            return "None";
+        }
+    }
+#endif // CHIP_DETAIL_LOGGING
+
+    /** Return subject value described in spec */
+    uint64_t GetValue() const
+    {
+        if (mSubject.Is<PaseSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<PaseSubject>().kPasscodeId) << 48;
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return mSubject.Get<NodeSubject>().nodeId;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<GroupSubject>().groupId) << 48;
+        }
+        else
+        {
+            return std::numeric_limits<uint64_t>::max();
+        }
+    }
+
+    bool operator==(const ScopedSubject & that) const { return this->mSubject == that.mSubject; }
+
+private:
+    Variant<PaseSubject, NodeSubject, GroupSubject> mSubject;
+};
+
+/**
+ * A subject is a global unique identifier. It suite 2 purpose:
+ *
+ *  1. Identify an entity. operator== can be used to check if 2 subjects are identical.
+ *  2. Associate to ACL entries to grant privileges
+ */
+class Subject
+{
+public:
+    Subject() : mFabricIndex(kUndefinedFabricIndex) {}
+
+    template <typename T, class... Args>
+    constexpr explicit Subject(FabricIndex fabricIndex, InPlaceTemplateType<T>, Args &&... args) :
+        mFabricIndex(fabricIndex), mScopedSubject(InPlaceTemplate<T>, std::forward<Args>(args)...)
+    {}
+
+    template <typename T, typename... Args>
+    static Subject Create(FabricIndex fabricIndex, Args &&... args)
+    {
+        return Subject(fabricIndex, InPlaceTemplate<T>, std::forward<Args>(args)...);
+    }
+
+    static Subject CreatePaseSubject()
+    {
+        // Return a unique subject for all PASE sessions.
+        return Subject(kUndefinedFabricIndex, InPlaceTemplate<PaseSubject>);
+    }
+
+    FabricIndex GetFabricIndex() const { return mFabricIndex; }
+    const ScopedSubject & GetScopedSubject() const { return mScopedSubject; }
+
+    bool operator==(const Subject & that) const
+    {
+        return this->mFabricIndex == that.mFabricIndex && this->mScopedSubject == that.mScopedSubject;
+    }
+
+private:
+    FabricIndex mFabricIndex;
+    ScopedSubject mScopedSubject;
+
+    friend bool operator==(const Subject &, const OperationalNodeId &);
+};
+
+/**
+ * OperationalNodeId identifies an individual Node on a Fabric. It is a special type of Subject targeting to NodeSubject. It is
+ * interchangeable with the generic Subject type but uses less memory.
```

Comment:
A few comments:

* This identifies an individual node on a fabric from the point of view of some specific node, not in general.
* "targeting to NodeSubject" does not make sense.  What is that trying to say?
* In what sense is this interchangeable with the generic Subject type?  Is the idea that you can compare these to Subject and can create a Subject from one of these?  So you can store this instead of a Subject if you happen to have a Subject representing a node?

Reply:
Yes, the idea is that:

1. you can compare these to Subject
2. you can create a Subject from OperationalNodeId ?
3. you can store this instead of a Subject if you happen to have a Subject representing a node ?